Source/JavaScriptCore/wasm/WasmBBQJIT.cpp

```diff
@@ -5927,39 +5985,312 @@ class BBQJIT {
+        result.setTryInfo(m_callSiteIndex, m_callSiteIndex, m_tryCatchDepth);
+        currentControlData().endBlock(*this, result, enclosingStack, true, false);
+
+        LOG_INSTRUCTION("Try", *signature);
+        LOG_INDENT();
+        splitStack(signature, enclosingStack, newStack);
+        result.startBlock(*this, newStack);
+        return { };
+    }
+
+    void emitCatchPrologue()
+    {
+        JIT_COMMENT(m_jit, "shared catch prologue");
+        m_jit.loadPtr(CCallHelpers::addressFor(CallFrameSlot::callee), GPRInfo::regT0);
+        m_jit.move(GPRInfo::regT0, GPRInfo::regT3);
+        m_jit.and64(CCallHelpers::TrustedImm64(JSValue::WasmMask), GPRInfo::regT3);
+        auto isWasmCallee = m_jit.branch64(CCallHelpers::Equal, GPRInfo::regT3, CCallHelpers::TrustedImm32(JSValue::WasmTag));
+        CCallHelpers::JumpList doneCases;
+        {
+            // FIXME: Handling precise allocations in WasmB3IRGenerator catch entrypoints might be unnecessary
+            // https://bugs.webkit.org/show_bug.cgi?id=231213
+            auto preciseAllocationCase = m_jit.branchTestPtr(CCallHelpers::NonZero, GPRInfo::regT0, CCallHelpers::TrustedImm32(PreciseAllocation::halfAlignment));
+            m_jit.andPtr(CCallHelpers::TrustedImmPtr(MarkedBlock::blockMask), GPRInfo::regT0);
+            m_jit.loadPtr(CCallHelpers::Address(GPRInfo::regT0, MarkedBlock::offsetOfHeader + MarkedBlock::Header::offsetOfVM()), GPRInfo::regT0);
+            doneCases.append(m_jit.jump());
+
+            preciseAllocationCase.link(&m_jit);
+            m_jit.loadPtr(CCallHelpers::Address(GPRInfo::regT0, PreciseAllocation::offsetOfWeakSet() + WeakSet::offsetOfVM() - PreciseAllocation::headerSize()), GPRInfo::regT0);
+            doneCases.append(m_jit.jump());
+        }
+
+        isWasmCallee.link(&m_jit);
+        m_jit.loadPtr(CCallHelpers::addressFor(CallFrameSlot::codeBlock), GPRInfo::regT0);
+        m_jit.loadPtr(CCallHelpers::Address(GPRInfo::regT0, Instance::offsetOfVM()), GPRInfo::regT0);
+
+        doneCases.link(&m_jit);
+
+        JIT_COMMENT(m_jit, "restore callee saves from vm entry buffer");
+        m_jit.restoreCalleeSavesFromVMEntryFrameCalleeSavesBuffer(GPRInfo::regT0, GPRInfo::regT3);
+
+        m_jit.loadPtr(CCallHelpers::Address(GPRInfo::regT0, VM::callFrameForCatchOffset()), GPRInfo::callFrameRegister);
+        m_jit.storePtr(CCallHelpers::TrustedImmPtr(nullptr), CCallHelpers::Address(GPRInfo::regT0, VM::callFrameForCatchOffset()));
+
+        JIT_COMMENT(m_jit, "Configure wasm context instance");
+        m_jit.loadPtr(CCallHelpers::Address(GPRInfo::callFrameRegister, CallFrameSlot::codeBlock * sizeof(Register)), GPRInfo::wasmContextInstancePointer);
+        m_frameSizeLabels.append(m_jit.moveWithPatch(TrustedImmPtr(nullptr), GPRInfo::regT0));
```

Comment:
Is there a reason we need to patch catch-block prologues? Maybe we can emit these at the end once we know `m_frameSize`?

Reply:
No reason right now. Just a quick implementation to make it work first. My plan is doing refactoring etc. after all the features are implemented and once we enabled & start testing. Until that, I tend to make it as simple as possible.